internal/backend/memory.go

```diff
@@ -8,117 +8,216 @@ import (
-func (s *StorageMemory) GetBucket(name string) error {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+func (s *StorageMemory) GetBucket(name string) (Bucket, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	bucketInMemory, err := s.getBucketInMemory(name)
+	return Bucket{bucketInMemory.Name, bucketInMemory.VersioningEnabled}, err
+}
 
-	if _, ok := s.buckets[name]; !ok {
-		return fmt.Errorf("no bucket named %s", name)
+func (s *StorageMemory) getBucketInMemory(name string) (bucketInMemory, error) {
+	if bucketInMemory, found := s.buckets[name]; found {
+		return bucketInMemory, nil
 	}
-	return nil
+	return bucketInMemory{}, fmt.Errorf("no bucket named %s", name)
 }
 
 // CreateObject stores an object
 func (s *StorageMemory) CreateObject(obj Object) error {
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
-
-	index := s.findObject(obj)
-	if index < 0 {
-		s.buckets[obj.BucketName] = append(s.buckets[obj.BucketName], obj)
-	} else {
-		s.buckets[obj.BucketName][index] = obj
+	bucketInMemory, err := s.getBucketInMemory(obj.BucketName)
+	if err != nil {
+		bucketInMemory = newBucketInMemory(obj.BucketName, false)
 	}
+	bucketInMemory.addObject(obj)
+	s.buckets[obj.BucketName] = bucketInMemory
 	return nil
 }
 
-// findObject looks for an object in its bucket and return the index where it
-// was found, or -1 if the object doesn't exist.
-//
-// It doesn't lock the mutex, callers must lock the mutex before calling this
-// method.
-func (s *StorageMemory) findObject(obj Object) int {
-	for i, o := range s.buckets[obj.BucketName] {
-		if obj.ID() == o.ID() {
-			return i
-		}
-	}
-	return -1
-}
-
 // ListObjects lists the objects in a given bucket with a given prefix and delimeter
 func (s *StorageMemory) ListObjects(bucketName string) ([]Object, error) {
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
-	objects, ok := s.buckets[bucketName]
-	if !ok {
-		return nil, errors.New("bucket not found")
+	bucketInMemory, err := s.getBucketInMemory(bucketName)
+	if err != nil {
+		return []Object{}, err
 	}
-	return objects, nil
+	return bucketInMemory.activeObjects, nil
 }
 
 // GetObject get an object by bucket and name
 func (s *StorageMemory) GetObject(bucketName, objectName string) (Object, error) {
-	obj := Object{BucketName: bucketName, Name: objectName}
+	return s.GetObjectWithGeneration(bucketName, objectName, 0)
+}
+
+// GetObjectWithGeneration retrieves an specific version of the object
+func (s *StorageMemory) GetObjectWithGeneration(bucketName, objectName string, generation int64) (Object, error) {
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
-	index := s.findObject(obj)
+	bucketInMemory, err := s.getBucketInMemory(bucketName)
+	if err != nil {
+		return Object{}, err
+	}
+	matchGeneration := false
+	obj := Object{BucketName: bucketName, Name: objectName}
+	listToConsider := bucketInMemory.activeObjects
+	if generation != 0 {
+		matchGeneration = true
+		obj.Generation = generation
+		listToConsider = append(listToConsider, bucketInMemory.archivedObjects...)
```

Comment:
I wonder if this can become too bad as ``archivedObjects`` grow. Not something to worry about now though :)

Hopefully it will never grow too much. We can adjust in the future if an issue is reported (or something like that)

Reply:
Its a very good point, but I think this is not a concern: looks the "unpack" with three dots does not actually extract each of the items and passes / manages them individually, but keeps the reference to the original archivedObjects slice (there are some interesting comments here: https://stackoverflow.com/questions/16248241/concatenate-two-slices-in-go ). In any case, an implementation with two searches, one in each slice, would be probably more performant (appending the slices has a cost in any case, as it may need to create a second slice to merge them), but in this specific scenario, I think its better to minimize duplication / keep current code readability rather than the performance win. Please, feel free to reopen if you think is worth changing that now